### PR TITLE
Copy between workspaces, rename time on copy

### PR DESCRIPTION
**Pull request by c4ffein**

### CODE REVIEW

**Review by vxgmichel — CHANGES_REQUESTED** [review]
Could you solve the conflicts so I can review the changes?

File: parsec/core/gui/files_widget.py [hunk 1]
```diff
@@ -328,24 +320,55 @@ def on_paste_clicked(self):
...
             while True:
                 try:
                     dst = self.current_directory / file_name
-                    if self.clipboard.status == Clipboard.Status.Cut:
-                        self.jobs_ctx.run(self.workspace_fs.move, src, dst)
+                    source_workspace = (
+                        self.clipboard.workspace if self.clipboard.workspace is not self else None
+                    )
+                    if self.clipboard.status == self.clipboard.Status.Cut:
+                        self.jobs_ctx.run(
```
Comment (vxgmichel) [inline parsec/core/gui/files_widget.py:332]: Isn't it time to move this while loop to a single job instead of looping over repeated jobs?

Comment (c4ffein) [reply parsec/core/gui/files_widget.py:332]: I don't know, I just adapted the existing code to enable that kind of copy, I don't know PyQt best practices, this is Max's code I guess... @Max-7 wdyt?

**Comment by bitlogik** [comment]
This doesn't allow the copy-paste of a directory between different workspace. It says "Cannot paste this file here" when pasting a directory.
It is a bug, an expected feature, a work in progress, planned in the future? Or this feature is only to copy paste files? If so, it should disable "Paste" contextual menu in the other workspace when a directory has been copied.

Here's the console during paste of a "/icon" directory in root "/" of an other ws (this PR adapted in our fork) :
```
2020-09-04 11:53:25 [exception] Unhandled error while cut/copy file [parsec.core.gui.files_widget] 
Traceback (most recent call last):
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 94, in _entry_id_from_path
    entry_id = manifest.children[name]
KeyError: 'icons'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/myuser/parsec-cloud/parsec/core/gui/files_widget.py", line 349, in on_paste_clicked
    source_workspace,
  File "/home/myuser/parsec-cloud/parsec/core/gui/trio_thread.py", line 229, in run
    return self._run_job(afn, *args)
  File "/home/myuser/parsec-cloud/parsec/core/gui/trio_thread.py", line 193, in _run_job
    return trio.from_thread.run(job, *args, trio_token=self._trio_token)
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 391, in from_thread_run
    return _run_fn_as_system_task(callback, afn, *args, trio_token=trio_token)
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 346, in _run_fn_as_system_task
    return q.get().unwrap()
  File "/home/myuser/.local/lib/python3.6/site-packages/outcome/_sync.py", line 111, in unwrap
    raise captured_error
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 384, in unprotected_afn
    return await afn(*args)
  File "/home/myuser/parsec-cloud/parsec/core/gui/files_widget.py", line 311, in workspace_fs_run_helper
    return await method(src, dst, source_workspace=source_workspace)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 464, in copytree
    source_files = await self.listdir(source_path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 289, in listdir
    return [child async for child in self.iterdir(path)]
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 289, in <listcomp>
    return [child async for child in self.iterdir(path)]
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 278, in iterdir
    info = await self.transactions.entry_info(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 176, in entry_info
    manifest, confined = await self._get_manifest_from_path(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 110, in _get_manifest_from_path
    entry_id, confined = await self._entry_id_from_path(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 96, in _entry_id_from_path
    raise FSFileNotFoundError(filename=path)
parsec.core.fs.exceptions.FSFileNotFoundError: No such file or directory: /icons/tray_icons
Traceback (most recent call last):
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 94, in _entry_id_from_path
    entry_id = manifest.children[name]
KeyError: 'icons'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/myuser/parsec-cloud/parsec/core/gui/files_widget.py", line 349, in on_paste_clicked
    source_workspace,
  File "/home/myuser/parsec-cloud/parsec/core/gui/trio_thread.py", line 229, in run
    return self._run_job(afn, *args)
  File "/home/myuser/parsec-cloud/parsec/core/gui/trio_thread.py", line 193, in _run_job
    return trio.from_thread.run(job, *args, trio_token=self._trio_token)
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 391, in from_thread_run
    return _run_fn_as_system_task(callback, afn, *args, trio_token=trio_token)
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 346, in _run_fn_as_system_task
    return q.get().unwrap()
  File "/home/myuser/.local/lib/python3.6/site-packages/outcome/_sync.py", line 111, in unwrap
    raise captured_error
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 384, in unprotected_afn
    return await afn(*args)
  File "/home/myuser/parsec-cloud/parsec/core/gui/files_widget.py", line 311, in workspace_fs_run_helper
    return await method(src, dst, source_workspace=source_workspace)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 464, in copytree
    source_files = await self.listdir(source_path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 289, in listdir
    return [child async for child in self.iterdir(path)]
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 289, in <listcomp>
    return [child async for child in self.iterdir(path)]
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 278, in iterdir
    info = await self.transactions.entry_info(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 176, in entry_info
    manifest, confined = await self._get_manifest_from_path(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 110, in _get_manifest_from_path
    entry_id, confined = await self._entry_id_from_path(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 96, in _entry_id_from_path
    raise FSFileNotFoundError(filename=path)
parsec.core.fs.exceptions.FSFileNotFoundError: No such file or directory: /icons/tray_icons
```

**Comment by bitlogik** [comment]
Contrary to the message printed "_Unhandled error while cut/copy file_", the error actually happens while pasting the directory.

**Comment by bitlogik** [comment]
I'm currently reviewing carefully our merge to see if there's any deviation from this PR.

**Comment by bitlogik** [comment]
Checked with this repo and this PR branch, I can confirm this issue when pasting a directory.

```
2020-09-04 13:08:21 [exception] Unhandled error while cut/copy file [parsec.core.gui.files_widget] 
Traceback (most recent call last):
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 86, in _lock_manifest_from_path
    entry_id = cast(LocalFolderishManifests, manifest).children[name]
KeyError: 'icons'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/myuser/parsec-cloud/parsec/core/gui/files_widget.py", line 346, in on_paste_clicked
    source_workspace,
  File "/home/myuser/parsec-cloud/parsec/core/gui/trio_thread.py", line 228, in run
    return self._run_job(afn, *args)
  File "/home/myuser/parsec-cloud/parsec/core/gui/trio_thread.py", line 192, in _run_job
    return trio.from_thread.run(job, *args, trio_token=self._trio_token)
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 391, in from_thread_run
    return _run_fn_as_system_task(callback, afn, *args, trio_token=trio_token)
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 346, in _run_fn_as_system_task
    return q.get().unwrap()
  File "/home/myuser/.local/lib/python3.6/site-packages/outcome/_sync.py", line 111, in unwrap
    raise captured_error
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 384, in unprotected_afn
    return await afn(*args)
  File "/home/myuser/parsec-cloud/parsec/core/gui/files_widget.py", line 308, in workspace_fs_run_helper
    return await method(src, dst, source_workspace=source_workspace)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 422, in copytree
    source_files = await self.listdir(source_path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 261, in listdir
    return [child async for child in self.iterdir(path)]
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 261, in <listcomp>
    return [child async for child in self.iterdir(path)]
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 250, in iterdir
    info = await self.transactions.entry_info(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 160, in entry_info
    manifest = await self._get_manifest_from_path(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 95, in _get_manifest_from_path
    async with self._lock_manifest_from_path(path) as manifest:
  File "/home/myuser/.local/lib/python3.6/site-packages/async_generator/_util.py", line 34, in __aenter__
    return await self._agen.asend(None)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 88, in _lock_manifest_from_path
    raise FSFileNotFoundError(filename=path)
parsec.core.fs.exceptions.FSFileNotFoundError: No such file or directory: /icons
Traceback (most recent call last):
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 86, in _lock_manifest_from_path
    entry_id = cast(LocalFolderishManifests, manifest).children[name]
KeyError: 'icons'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/myuser/parsec-cloud/parsec/core/gui/files_widget.py", line 346, in on_paste_clicked
    source_workspace,
  File "/home/myuser/parsec-cloud/parsec/core/gui/trio_thread.py", line 228, in run
    return self._run_job(afn, *args)
  File "/home/myuser/parsec-cloud/parsec/core/gui/trio_thread.py", line 192, in _run_job
    return trio.from_thread.run(job, *args, trio_token=self._trio_token)
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 391, in from_thread_run
    return _run_fn_as_system_task(callback, afn, *args, trio_token=trio_token)
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 346, in _run_fn_as_system_task
    return q.get().unwrap()
  File "/home/myuser/.local/lib/python3.6/site-packages/outcome/_sync.py", line 111, in unwrap
    raise captured_error
  File "/home/myuser/.local/lib/python3.6/site-packages/trio/_threads.py", line 384, in unprotected_afn
    return await afn(*args)
  File "/home/myuser/parsec-cloud/parsec/core/gui/files_widget.py", line 308, in workspace_fs_run_helper
    return await method(src, dst, source_workspace=source_workspace)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 422, in copytree
    source_files = await self.listdir(source_path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 261, in listdir
    return [child async for child in self.iterdir(path)]
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 261, in <listcomp>
    return [child async for child in self.iterdir(path)]
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/workspacefs.py", line 250, in iterdir
    info = await self.transactions.entry_info(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 160, in entry_info
    manifest = await self._get_manifest_from_path(path)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 95, in _get_manifest_from_path
    async with self._lock_manifest_from_path(path) as manifest:
  File "/home/myuser/.local/lib/python3.6/site-packages/async_generator/_util.py", line 34, in __aenter__
    return await self._agen.asend(None)
  File "/home/myuser/parsec-cloud/parsec/core/fs/workspacefs/entry_transactions.py", line 88, in _lock_manifest_from_path
    raise FSFileNotFoundError(filename=path)
parsec.core.fs.exceptions.FSFileNotFoundError: No such file or directory: /icons

```

**Comment by Max-7** [comment]
Done by #1540. Closing this PR.

**Pull request closed without merging**